File: app/src/main/java/com/example/messenger/RegistrationViewModel.java
package com.example.messenger;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

public class RegistrationViewModel extends ViewModel {

    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final MutableLiveData<FirebaseUser> user = new MutableLiveData<>();

    private final FirebaseAuth auth;
    private final FirebaseDatabase database;
    private final DatabaseReference userReference;

    public RegistrationViewModel() {
        auth = FirebaseAuth.getInstance();
        auth.addAuthStateListener(userFB -> {
            if (userFB.getCurrentUser() != null) {
                user.setValue(userFB.getCurrentUser());
            }
        });
        database = FirebaseDatabase.getInstance();
        userReference = database.getReference("Users");
    }

    public LiveData<String> getError() {
        return error;
    }

    public LiveData<FirebaseUser> getUser() {
        return user;
    }

    public void register(String email, String password, String name, String lastName, int age) {
        auth.createUserWithEmailAndPassword(email, password)
                .addOnSuccessListener(authResult -> {
                    FirebaseUser firebaseUser = authResult.getUser();
                    if(firebaseUser == null) {
                        return;
                    }
                    String uid = firebaseUser.getUid();
                    User user = new User(
                            uid,
                            name,
                            lastName,
                            age,
                            false
                    );
                    userReference.child(uid).setValue(user);
                })
                .addOnFailureListener(e -> error.setValue(e.getMessage()));
    }
}

File: app/src/main/java/com/example/messenger/RegistrationViewModel.kt
package com.example.messenger

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.messenger.pojo.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.FirebaseDatabase

class RegistrationViewModel : ViewModel() {

    private val error = MutableLiveData<String>()
    private val user = MutableLiveData<FirebaseUser>()

    private val auth = FirebaseAuth.getInstance()
    private val database = FirebaseDatabase.getInstance()
    private val userReference = database.getReference("Users")

    init {
        auth.addAuthStateListener { it.currentUser?.let { user.value = it } }
    }

    val getError: LiveData<String> get() = error
    val getUser: LiveData<FirebaseUser> get() = user

    fun register(email: String, password: String, name: String, lastName: String, age: Int) {
        auth.createUserWithEmailAndPassword(email, password)
                .addOnSuccessListener {
                    val firebaseUser = it.user
                    val uid = firebaseUser?.uid ?: return@addOnSuccessListener
                    val user = User(uid, name, lastName, age, false)
                    userReference.child(uid).setValue(user)
                }
                .addOnFailureListener { error.value = it.message }
    }

}